enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/IdentityVerificationOtpService.java

```diff
@@ -133,6 +164,66 @@ public boolean isUserVerifiedUsingOtp(String processId) {
         return otp.getStatus() == OtpStatus.VERIFIED;
     }
 
+    private OtpVerifyResponse verifyPresenceCheck(final OnboardingProcessEntity process, final OtpVerifyResponse response) throws OnboardingProcessException {
+        final String processId = process.getId();
+        if (!identityVerificationConfig.isPresenceCheckEnabled()) {
+            logger.debug("Presence check is not enabled, Process ID: {}", processId);
+            return response;
+        }
+
+        logger.debug("Evaluating result of presence check, Process ID: {}", processId);
+
+        final IdentityVerificationEntity idVerification = getIdentityVerificationEntity(process);
+        final String errorDetail = idVerification.getErrorDetail();
+        final String rejectReason = idVerification.getRejectReason();
+
+        if (StringUtils.isNotBlank(errorDetail) || StringUtils.isNotBlank(rejectReason)) {
+            logger.info("SCA failed, IdentityVerification ID: {} of Process ID: {} contains errorDetail: {}, rejectReason: {} from previous step",
+                    idVerification.getId(), processId, errorDetail, rejectReason);
+            return moveToPhasePresenceCheck(process, response, idVerification);
+        } else {
+            logger.debug("PRESENCE_CHECK without error or reject reason, Process ID: {}", idVerification.getProcessId());
+        }
+        return response;
+    }
+
+    private IdentityVerificationEntity getIdentityVerificationEntity(final OnboardingProcessEntity process) throws OnboardingProcessException {
+        return identityVerificationRepository.findFirstByActivationIdOrderByTimestampCreatedDesc(process.getActivationId()).orElseThrow(() ->
+                new OnboardingProcessException("No IdentityVerification found for Process ID: " + process.getId()));
+    }
+
+    private OtpVerifyResponse moveToPhasePresenceCheck(
```

Comment:
Hopefully, it will be refactored later on, when it is rewritten to the state machine. I guess to publish an event or so.